text/0000-feature-metadata-doc-cfg.md

```diff
@@ -0,0 +1,334 @@
+
+1. Add a way to write feature descriptions in `Cargo.toml`, as well some other
+   simple feature attributes.
+2. Establish a way for Cargo or other build systems to easily pass advanced
+   configuration information to `rustdoc` (e.g., favicon or playground URLs from
+   `Cargo.toml`)
+3. Allow `rustdoc` to accept this information either from the CLI or from a new
+   `rustdoc.toml` file.
+
+The outcome is that `rustdoc` will gain the ability to document cargo features
+and get its configuration from `Cargo.toml`, without having any awareness of
+Cargo project structure itself. There will also be room to grow for more
+advanced configuration options.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, `docs.rs` provides a basic view of available feature flags on a
+rather simple page: for example, [`tokio`]. It is helpful as a quick overview of
+available features, but it is not managed by `rustdoc` (i.e., is not available
+on local) and there is no way to specify a description or other useful
+information.
+
+The second problem is that `rustdoc` has some per-crate configuration settings,
+such as relevant URLs, that are awkward to define in Rust source files using
+attributes. It is expected that there may be further configuration options in
+the future, for specifying things like:
+
+1. Resource manifests (paths to assets, such as `KaTeX` for math rendering or
+   non-image files)
+2. Non-code informational pages (such as [`clap`'s derive information])
+
+This RFC provides a way to solve both problems: it specifies a way to add
+user-facing metadata to cargo features, and specifies how that and other
+information should be passed to `rustdoc`.
+
+[`tokio`]: https://docs.rs/crate/tokio/latest/features
+[`clap`'s derive information]: https://docs.rs/clap/4.2.2/clap/_derive/index.html
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Usage is simple: features will be able to be specified in a table (inline or
+separate) with the keys `doc`, `public`, `deprecated`, and `requires`. Sample
+`Cargo.toml`:
+
+```toml
+# Cargo.toml
+
+[features]
+# current configuration will continue to work
+foo = []
+# Add a description to the feature. Equivalent to today's `bar = ["foo"]`
+bar = { requires = ["foo"], doc = "simple docstring here"}
+# `public` indicates whether or not the feature should be public in
+# documentation and usable by downstream users; defaults to `true`.
+baz = { requires = ["foo"], public = false, deprecated = true }
+
+# Features can also be full tables if descriptions are longer
+[features.qux]
+requires = ["bar", "baz"]
+doc = """
+# qux
+
+This could be a longer description of this feature
+"""
+```
+
+This RFC will also allow a `[tools.rustdoc]` table where existing configuration
+can be specified:
+
+```toml
+# Cargo.toml
+
+[tools.rustdoc]
+html-logo-url = "https://example.com/logo.jpg"
+issue-tracker-base-url = "https://github.com/rust-lang/rust/issues/"
```

Comment:
I feel like also adding a `tools` table is overoading this RFC with too many things.

Note #3389 might also be related

Reply:
@est31 did mention that here https://github.com/rust-lang/rfcs/pull/3416#discussion_r1167369030

I don't think I'm seeing it in that RFC, but how is clippy made aware of the contents of the `[lints]` table? I see a JSON schema mentined a couple times in the thread.

I was loosely imagining that there could eventually be a simple schema defined for any tool, official or inofficial. Basically the tool could specify which tables from Cargo.toml cargo should pass it, making it easy to run any tool without cargo itself being "aware" of its configuration (and non-cargo build systems could easily use the same patterns). Not that I want to include that in this RFC, but the rough idea is included in future possibilities

```json5
{   // rustdoc config
    "tool": "rustdoc",
    "top-level-tables": ["tools.rustdoc"],  // serialized as top-level when passed to tool
    "nested-tables": ["features"],          // serialized as an object with the same name when passed
    "json-arg-name": "--config-json",       // how to pass data from Cargo.toml
    "config-file": ["rustdoc.toml"],        // files that cargo should search for
    "config-arg-name": "--config-file"      // how to pass file paths
    "accepts-spanned-json": true            // whether cargo should include spans from Cargo.toml
},
{   // clippy config.
    "tool": "clippy",
    "top-level-tables": ["tools.clippy"],
    "nested-tables": ["lints"],
    // ...
}
// ... any tools just needs
```

Reply:
(this is also being discussed on zulip)

> I don't think I'm seeing it in that RFC, but how is clippy made aware of the contents of the [lints] table? I see a JSON schema mentined a couple times in the thread.

In its current form, the RFC only covers passing in lint levels which does not need a JSON schema.  There are two cases where this might be needed in the future
- Tool/lint configuration but `cargo clippy` might handle reading it and passing it down, so its not my problem :)
- Auto-sort lint levels.  Like I said, this is for the future, so not worrying about it too much.

Reply:
The RFC has been split, this is now at https://github.com/rust-lang/rfcs/pull/3421